src/backend/commands/tablespace.c

```diff
@@ -1649,26 +1650,39 @@ assign_temp_tablespaces(const char *newval, void *extra)
 		SetTempTablespaces(NULL, 0);
 }
 
+/* assign_hook: do extra actions as needed */
+void
+assign_file_temp_tablespaces(const char *newval, void *extra)
+{
+	temp_tablespaces_extra *myextra = (temp_tablespaces_extra *) extra;
+
+	/*
+	 * If check_temp_tablespaces was executed inside a transaction, then pass
+	 * the list it made to fd.c.  Otherwise, clear fd.c's list; we must be
+	 * still outside a transaction, or else restoring during transaction exit,
+	 * and in either case we can just let the next PrepareTempTablespaces call
+	 * make things sane.
+	 */
+	if (myextra)
+		SetFileTempTablespaces(myextra->tblSpcs, myextra->numSpcs);
+	else
+		SetFileTempTablespaces(NULL, 0);
+}
+
 /*
- * PrepareTempTablespaces -- prepare to use temp tablespaces
- *
  * If we have not already done so in the current transaction, parse the
  * temp_tablespaces GUC variable and tell fd.c which tablespace(s) to use
- * for temp files.
+ * for temporary files or tables.
  */
-void
-PrepareTempTablespaces(void)
+static void
+PrepareTablespacesImpl(char *str, void (*set)(Oid *, int))
```

Comment:
Maybe give the arguments of this function `str` and `set` more beautiful and understandable names? Let's say `tablespaces` and `SetTablespaces`